homeassistant/components/vesync/humidifier.py

```diff
@@ -0,0 +1,155 @@
+    """Set up the VeSync humidifier platform."""
+
+    async def async_discover(devices):
+        """Add new devices to platform."""
+        _async_setup_entities(devices, async_add_entities)
+
+    disp = async_dispatcher_connect(
+        hass, VS_DISCOVERY.format(VS_HUMIDIFIERS), async_discover
+    )
+    hass.data[DOMAIN][VS_DISPATCHERS].append(disp)
+
+    _async_setup_entities(hass.data[DOMAIN][VS_HUMIDIFIERS], async_add_entities)
+
+
+@callback
+def _async_setup_entities(devices, async_add_entities):
+    """Check if device is online and add entity."""
+    dev_list = []
+    for dev in devices:
+        if DEV_TYPE_TO_HA.get(dev.device_type) == "humidifier":
+            dev_list.append(VeSyncHumidifierHA(dev))
+        else:
+            continue
+
+    async_add_entities(dev_list, update_before_add=True)
+
+
+class VeSyncHumidifierHA(VeSyncToggleDevice, HumidifierEntity):
+    """Representation of a VeSync humidifier."""
+
+    _attr_max_humidity = MAX_HUMIDITY
+    _attr_min_humidity = MIN_HUMIDITY
+
+    def __init__(self, device):
+        """Initialize the VeSync humidifier device."""
+        super().__init__(device)
+        self._attr_available_modes = []
+        self._attr_available_modes = AVAILABLE_MODES[self.device.device_type]
+
+    @property
+    def supported_features(self):
+        """Flag supported features."""
+        return SUPPORT_MODES
+
+    @property
+    def target_humidity(self):
+        """Return the humidity we try to reach."""
+        return self.device.config["auto_target_humidity"]
+
+    @property
+    def mode(self):
+        """Get the current preset mode."""
+        if self.device.details["mode"] in self.available_modes:
+            return self.device.details["mode"]
+        return None
+
+    @property
+    def is_on(self):
+        """Return True if humidifier is on."""
+        return self.device.enabled  # device_status is always on
+
+    @property
+    def unique_info(self):
+        """Return the ID of this humidifier."""
+        return self.device.uuid
+
+    @property
+    def extra_state_attributes(self):
+        """Return the state attributes of the humidifier."""
+        attr = {}
+
+        if "water_lacks" in self.device.details:
+            attr["water_lacks"] = self.device.details["water_lacks"]
+
+        if "humidity_high" in self.device.details:
+            attr["humidity_high"] = self.device.details["humidity_high"]
+
+        if "water_tank_lifted" in self.device.details:
+            attr["water_tank_lifted"] = self.device.details["water_tank_lifted"]
+
+        if "automatic_stop_reach_target" in self.device.details:
+            attr["automatic_stop_reach_target"] = self.device.details[
+                "automatic_stop_reach_target"
+            ]
+
+        if "humidity" in self.device.details:
+            # The "humidity" key is reserved by HumidifierEntity for target humidity
+            attr["current_humidity"] = self.device.details["humidity"]
```

Comment:
Consider exposing this as a separate sensor instead of an attribute. My PR #61837 adds device registrations for devices in this component, which would allow you to tie together the humidifier and the humidity sensor into a single device.

Reply:
I'm pretty new to writing Home Assistant integrations. Are entities always considered strictly better than state attributes that can be turned into entities using templates? Are there guidelines for deciding what should be an entity and what should be a state attribute? Instead of making less-commonly used properties state attributes, is it not better to make them entities tagged with `entity_registry_enabled_default=False`?

The reason I'm asking is that I wonder if I should turn the other state attributes into separate entities as well.

Reply:
I think about it in terms of use case -- how likely is it that someone will want to use the information? If it's just simple diagnostic data that's infrequently referenced, attributes make sense. But I feel like a humidity sensor is pretty useful top-level information to expose.

But I am no expert either -- only a handful of PRs merged so far 😉 I'd love to hear from the maintainers on this.

Reply:
Funny enough, diagnostic tools are precisely one of the examples provided of what should be an entity that's disabled by default (see https://developers.home-assistant.io/docs/core/entity/#advanced-properties). The guidance is certainly nebulous, so I'd love to hear from the maintainers as well.